docs/bugs.json

```
@@ -1,4 +1,11 @@
 [
+	{
+		"name": "KeccakCaching",
+		"summary": "The bytecode optimizer incorrectly evaluated certain keccak256 hashes. You are unlikely to be affected if you do not compute keccak256 hashes in inline assembly.",
+		"description": "Solidity's bytecode optimizer has a step that can compute keccak256 hashes, if the contents of the memory are known during compilation time. This step also had a mechanism to determine that two keccak256 hashes are equal even if the values in memory are not known during compile time. This mechanism had a bug where, keccak256 of the same memory location, but different sizes were considered equal. More specifically, ``keccak256(mpos, length1)`` and ``keccak256(mpos, length2)`` in some cases were considered equal when ``length1`` and ``length2``, when rounded up to nearest multiple of 32 were the same. You maybe affected if you compute keccak256 hashes from the same memory location, but different lengths in inline assembly and enabled the optimizer.",
```

```suggestion
		"description": "Solidity's bytecode optimizer has a step that can compute keccak256 hashes, if the contents of the memory are known during compilation time. This step also has a mechanism to determine that two keccak256 hashes are equal even if the values in memory are not known during compile time. This mechanism had a bug where, keccak256 of the same memory location, but different sizes were considered equal. More specifically, ``keccak256(mpos, length1)`` and ``keccak256(mpos, length2)`` in some cases were considered equal when ``length1`` and ``length2``, when rounded up to nearest multiple of 32 were the same. You maybe affected if you compute keccak256 hashes from the same memory location, but different lengths in inline assembly and enabled the optimizer.",
```
Plus: I don't think that the memory locations need to be the same, the optimizer only has to deduce if the contents are the same.

---

Here is an example:
```
contract C {
    function f() public view returns (bool ret) {
        assembly {
            let x := calldataload(0)
            mstore(0, x)
            mstore(0x20, x)
            let a := keccak256(0, 4)
            let b := keccak256(0x20, 8)
            ret := eq(a, b)
        }
    }
}
```